### PR TITLE
perf(admin-events): fix slow mutex in writer

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This PR fixes issues seen during CI where a slow mutex was detected in the administration handler when flushing the buffered events.

The main issue is that previously within the `Upsert` method of the writer a database lookup was done for the event unconditionally.

Since the frequency of `Upsert` being called is directly correlated to the log events being emitted, this put a) a huge amount of read operations on the database and b) lead to the `Upsert` methods holding / waiting on the lock for a long time.

Subsequently, when the `Flush` operation was called from the different go routine, this lead to waiting on the lock.

As a solution, the database lookup was moved to the `Flush` method instead, minimizing the read operations by using `GetMany` and only doing those when the buffer is full.

The benchmark results ran locally can be found below, with a comparison of the previous results:
```
goos: darwin
goarch: amd64
pkg: github.com/stackrox/rox/central/administration/events/datastore
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
BenchmarkDatastore_Add_and_Flush
BenchmarkDatastore_Add_and_Flush/add_1000_events_to_the_writer_and_flush_it
BenchmarkDatastore_Add_and_Flush/add_1000_events_to_the_writer_and_flush_it-12         	      50	  15665028 ns/op
BenchmarkDatastore_Add_and_Flush/add_5000_events_to_the_writer_and_flush_it
BenchmarkDatastore_Add_and_Flush/add_5000_events_to_the_writer_and_flush_it-12         	      50	 111444743 ns/op
BenchmarkDatastore_Add_and_Flush/add_10000_events_to_the_writer_and_flush_it
BenchmarkDatastore_Add_and_Flush/add_10000_events_to_the_writer_and_flush_it-12        	      50	 229537995 ns/op
BenchmarkDatastore_Add_and_Flush/add_20000_events_to_the_writer_and_flush_it
BenchmarkDatastore_Add_and_Flush/add_20000_events_to_the_writer_and_flush_it-12        	      50	 478561343 ns/op


goos: darwin
goarch: amd64
pkg: github.com/stackrox/rox/central/administration/events/datastore
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
                                                                       │    old.txt     │                new.txt                │
                                                                       │     sec/op     │    sec/op     vs base                 │
Datastore_Add_and_Flush/add_1000_events_to_the_writer_and_flush_it-12     836.92m ± ∞ ¹   15.67m ± ∞ ¹        ~ (p=1.000 n=1) ²
Datastore_Add_and_Flush/add_5000_events_to_the_writer_and_flush_it-12     6038.3m ± ∞ ¹   111.4m ± ∞ ¹        ~ (p=1.000 n=1) ²
Datastore_Add_and_Flush/add_10000_events_to_the_writer_and_flush_it-12   12902.2m ± ∞ ¹   229.5m ± ∞ ¹        ~ (p=1.000 n=1) ²
Datastore_Add_and_Flush/add_20000_events_to_the_writer_and_flush_it-12   27043.1m ± ∞ ¹   478.6m ± ∞ ¹        ~ (p=1.000 n=1) ²
geomean                                                                     6.480         117.7m        -98.18%
```

This is an overall increase of performance by over 90% percent.

## Checklist
- [ ] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

- see CI passing.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: central/administration/events/datastore/internal/writer/writer.go [hunk 1]
```diff
@@ -11,17 +11,16 @@ import (
 // Writer implements a buffered write for the administration events datastore.
 //
 // Since we generate events from logs, there could be many log events created
-// in a small time frame. The buffer is essentially a write optimization such
+// in a small-time frame. The buffer is essentially a write optimization such
```
Comment (stehessel) [inline central/administration/events/datastore/internal/writer/writer.go:14]: Why this change?

File: central/administration/events/datastore/internal/writer/writer.go [hunk 2]
```diff
@@ -11,17 +11,16 @@ import (
...
 // While event ingestion would probably be fine without the buffer from Central's
 // perspective, it would increase the load on the database due to many concurrent
 // transactions and table locks.
 //
 // When events are upserted to the writer, they first end up in a buffer.
-// The buffered event has the most recent event state. If an entry for
-// the event is already present in the data store, this record is merged
-// with the buffered record. The buffer is written to the data store once
-// the writer is flushed.
+// The buffered event has the most recent event state.The buffer is written to the
```
Comment (stehessel) [inline central/administration/events/datastore/internal/writer/writer.go:21]:
```suggestion
// The buffered event has the most recent event state. The buffer is written to the
```

File: central/administration/events/datastore/internal/writer/writer_impl.go [hunk 3]
```diff
@@ -37,16 +40,41 @@ func (c *writerImpl) resetNoLock() {
 }
 
 func (c *writerImpl) flushNoLock(ctx context.Context) error {
-	eventChunk := make([]*storage.AdministrationEvent, 0, len(c.buffer))
-	for _, event := range c.buffer {
-		eventChunk = append(eventChunk, event)
+	eventsToAdd := sliceutils.ShallowClone(maputil.Values(c.buffer))
+
+	// Short-circuit here in case we have no events to be added.
```
Comment (stehessel) [inline central/administration/events/datastore/internal/writer/writer_impl.go:45]: Can't we short circuit already first thing first in this function if `len(c.buffer) == 0`?

**Review by stehessel — APPROVED** [review]
Nice optimization 👍 . I think `perf` would fit better than `fix` in this case for the conventional commit. Or better yet just use the ROX ticket of the mutex flake.

File: central/administration/events/datastore/internal/writer/writer.go (hunk 1 above)
Comment (dhaus67) [reply central/administration/events/datastore/internal/writer/writer.go:14]: Goland's spell check 🤷

Comment (dhaus67) [reply central/administration/events/datastore/internal/writer/writer.go:14]: Yeah it's cause it should have been `small timeframe` from the start, so changed it to that.
